### ISSUE TITLE
Cumulocity registration is sent 2 times if an entity is auto-registered

### ISSUE THREAD

**Issue by Bravo555** [thin-edge/thin-edge.io#2311]
**Describe the bug**

Because the c8y mapper publishes an entity registration message when auto-registering an entity (so other clients can be aware of a new entity),  a cloud registration message can be sent 2 times. This is because when I was initially working on the entity store, I didn't realise that if an MQTT client subscribes to topic `T` and publishes a message on topic `T`, the broker will echo the message back to the client. This makes it so that when c8y mapper auto-registers an entity it publishes a registration message for other clients, but it will pick up this message again, leading to a 2nd `entity_store.update()` being called and 2nd `self.try_convert_entity_registration()` being called.

**To Reproduce**

1. Publish a message on a topic of entity that wasn't registered, i.e. `te/device/main/service/my_not_registered_service_1/m/my_measurement_series`:
   ```
   $ tedge mqtt pub te/device/main/service/my_not_registered_service_1/m/my_measurement_series '{"temperature": 21.37}'
   ```

2. The [`102 (service creation)`](https://cumulocity.com/guides/reference/smartrest-two/#service-creation-102) smartrest message has been sent twice:

    ```
    [te/device/main/service/my_not_registered_service_2/m/my_measurement_series] {"temperature": 21.37}
    [te/device/main/service/my_not_registered_service_2] {"@id":"devcontainer:device:main:service:my_not_registered_service_2","@parent":"device/main//","@type":"service","name":"my_not_registered_service_2","type":"systemd"}

    **[c8y/s/us] 102,devcontainer:device:main:service:my_not_registered_service_2,systemd,my_not_registered_service_2,up**
    [c8y/measurement/measurements/create] {"temperature":{"temperature":{"value":21.37}},"time":"2023-10-03T17:51:25.679036501Z","type":"my_measurement_series"}

    **[c8y/s/us] 102,devcontainer:device:main:service:my_not_registered_service_2,systemd,my_not_registered_service_2,up**
    ```

**Expected behavior**

Entity auto-registration leads only to a single cloud registration being sent.

**Additional context**

In itself, this bug doesn't break anything. However I submitted it, because it's much more annoying when an initial message which triggers auto-registration is not a measurement message, but a health message. A [service registration message](https://cumulocity.com/guides/reference/smartrest-two/#service-creation-102) contains `service status` field, which is set to `"up"` by default. When the initial message contains any other status, it's being overridden, e.g. when we publish a health status message with `status=unknown`, the following messages are sent:

```
**processing health message itself**
1. 102 from auto-registration, status=up
2. 102 from health, status=unknown

** we get a duplicated registration message**
3. 102 from registration message, status=up (INCORRECT)
```

**Proposed solutions**

As many subtle bugs can be generated by publishing MQTT messages on the same topic we subscribe from, we should look there for a solution. It might be possible to implement deduplication on an MQTT actor that works in a following way:

```
When a consumer publishes a message on a topic which is part of its subscribe filter, save a hash of the message (topic + payload) to an ignored messages set. For every incoming message, if its hash is in the ignored message set, don't pass the message to the consumer and remove the hash from the set.
```

This approach may be lacking if the same message being published again carries information derived from the time the message was published, e.g. health status messages published in response to a watchdog. However I don't expect much issues because:

1. in such cases usually topics these clients subscribe to and published on are different precisely to avoid feedback loops
2. for messages where we implicitly derive information from time when a message was received, we should instead explicitly put a timestamp in a message payload. Including timestamp in a health status message and other similar messages might also provide user value.

**Comment by Bravo555** [comment]
@albinsuresh this affects entity registration, so I'd like to hear your feedback on whether you agree that this behaviour (publishing 2 cloud registration messages) is undesirable, and any thoughts on possible solutions.

**Comment by albinsuresh** [comment]
Yes, this was one of the follow-up tasks planned [here](https://github.com/thin-edge/thin-edge.io/pull/2266#issuecomment-1735325392). For the fix, I wouldn't fix it at the MQTT actor layer, as a subscribing client publishing to the same topic, prompting it to reprocess it, may be valid in some other use-cases. So, to avoid this specific issue, we can just update the `try_convert_entity_registration` logic to avoid the re-registration and conversion of an incoming registration message if that entity already exists in the entity store with the same metadata provided in the message.

**Comment by Bravo555** [comment]
> we can just update the `try_convert_entity_registration` logic to avoid the re-registration and conversion of an incoming registration message if that entity already exists in the entity store with the same metadata provided in the message.

As service status is not currently part of entity metadata (although in the future it probably should be), it will be good workaround for now for health status messages. I'll make necessary changes as part of this PR.
